[FLINK-9337] Implemented AvroDeserializationSchema

dawidwys opened the pull request:
## What is the purpose of the change
Provides implementation of AvroDeserializationSchema that reads records serialized as avro and also provides version that uses Confluent Schema Registry to look up writer schema.


## Brief change log

  - Implemented AvroDeserializationSchema / RegistryAvroDeserializationSchema / ConfluentRegistryAvroDeserializationSchema
  - Extended AvroSerializer to handle GenericRecords
  - Added GenericRecordTypeInformation

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**yes** / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (**yes** / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

StephanEwen commented:
Thanks, the main code looks good!

Unfortunately, this seems to wither break the compatibility with prior savepoints (when Avro types were implicitly handled through Kryo, now bridged through the `BackwarsCompatibleAvroSerializer`) or needs to adjust that test.

There are also some license header issues, causing the build to fail.

dawidwys commented:
Right sorry for that. I've changed the data generator a bit, so it produced different results than before with the same seed. I've recreated the serialized data with updated `TestDataGenerator`. It took me a while though to figure out that it should be created with current code rather than 1.3 branch. Therefore I updated the comment accordingly.

Also reworded a bit other names as the `BackwardsCompatibleAvroSerializerTest` does not test compatibility with 1.3, but only with `PojoSerializer`.

twalthr commented on flink-formats/flink-avro-confluent-registry/pom.xml:
```diff
@@ -0,0 +1,94 @@
+	</parent>
+	<modelVersion>4.0.0</modelVersion>
+
+	<artifactId>flink-avro-confluent-registry</artifactId>
```
Do we really want the Confluent Schema Registry code to be in `flink-formats`? Shouldn't this be in something like `flink-catalogs` in the future?

StephanEwen replied:
That is a good question - maybe eventually. We could leave it in `flink-formats` for now until we have a case to create `flink-catalogs`.

This is also not a full-blown catalog support, as for the Table API, but something much simpler - just multiple Avro Schemas.

StephanEwen commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroDeserializationSchema.java:
```diff
@@ -0,0 +1,215 @@
+/*
```
This class uses a mixture of eager initialization of transient members (in readObject()) and lazy initialization (in getDatumReader()).

I would suggest to do it all one way or the other.

My suggestion would be to avoid `readObject()` whenever possible. If you encounter an exception during schema parsing (and it may be something weird from Jackson, like a missing manifest due to a shading issue), you will get the most unhelpful exception stack trace ever, in the weirdest place (like Flink's RPC message decoder).

In my experience, when a user sees such a stack trace, they are rarely able to diagnose that. Best case they show up on the mailing list, worst case they give up.

StephanEwen replied:
That would mean initializing it all lazily, in `getDatumReader()` or in a `checkAvroInitialized()` method.

StephanEwen commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroDeserializationSchema.java:
```diff
@@ -0,0 +1,215 @@
+		}
+
+		if (SpecificRecord.class.isAssignableFrom(recordClazz)) {
+			this.reader = SpecificData.get().getSchema(recordClazz);
```
Do we need to pass the use code class loader into the SpecificData / GenericData / etc? In order to resolve nested user-defined types?

dawidwys replied:
Didn't think about it, sorry.

StephanEwen commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/RegistryAvroDeserializationSchema.java:
```diff
@@ -0,0 +1,82 @@
+ * @param <T> type of record it produces
+ */
+public class RegistryAvroDeserializationSchema<T> extends AvroDeserializationSchema<T> {
+	private final SchemaCoder.SchemaCoderProvider schemaCoderProvider;
```
We typically do empty lines between class declaration and members.

StephanEwen commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/RegistryAvroDeserializationSchema.java:
```diff
@@ -0,0 +1,82 @@
+		Class<T> recordClazz,
+		@Nullable Schema reader,
+		SchemaCoder.SchemaCoderProvider schemaCoderProvider) {
+		super(recordClazz, reader);
```
Empty line / indentation

StephanEwen commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/RegistryAvroDeserializationSchema.java:
```diff
@@ -0,0 +1,82 @@
+		}
+	}
+
+	private void readObject(ObjectInputStream ois) throws ClassNotFoundException, IOException {
```
See above, would suggest to avoid readObject()

StephanEwen commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSerializer.java:
```diff
@@ -79,15 +87,16 @@
+	private transient GenericData avroData;
 
 	private transient Schema schema;
+	private final String schemaString;
```
As per the comments, the existing code orders config fields before runtime fields. Can you place the schema to match that pattern?

StephanEwen commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSerializer.java:
```diff
@@ -99,9 +108,29 @@
+	 * For serializing {@link GenericData.Record} use {@link AvroSerializer#AvroSerializer(Class, Schema)}
 	 */
 	public AvroSerializer(Class<T> type) {
+		Preconditions.checkArgument(!isGenericRecord(type),
```
Minor: Other preconditions checks in this class are done by statically imported methods. While this is not consistent within the code base, I would suggest to keep this consistent within a class as much as possible.

StephanEwen commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSerializer.java:
```diff
@@ -275,9 +304,19 @@ else if (configSnapshot instanceof AvroSerializerConfigSnapshot) {
 	public TypeSerializer<T> duplicate() {
-		return new AvroSerializer<>(type);
+		if (schemaString != null) {
+			return new AvroSerializer<>(type, new Schema.Parser().parse(schemaString));
```
Duplication happens frequently, would be good to avoid schema parsing. You can add a private copy constructor that takes class and string.

dawidwys replied:
Didn't think it through well. Thought we need to create a deep copy of the schema, but as it is stateless I think we can just pass the schema. My mistake. Correct me if I am wrong.

dawidwys commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroDeserializationSchema.java:
```diff
@@ -0,0 +1,215 @@
+	@SuppressWarnings("unchecked")
+	public TypeInformation<T> getProducedType() {
+		if (SpecificRecord.class.isAssignableFrom(recordClazz)) {
+			return new AvroTypeInfo(recordClazz, false);
```
Agree raw types should be avoided. Unfortunately I cannot find a way to avoid it here. Your solution does not work cause the `recordClazz.asSubclass(..)` returns `<? extends SpecificRecord>` which does not conforms to `T` of the returned type.

StephanEwen replied:
Oh, I see. I guess you can push the type info creation into a static method with a different generic scope, but that would not be simpler to read. Would keep it like it is then.

dawidwys commented:
As for the snapshot binary data, I do understand that it should be created with appropriate flink version (in this case in theory with flink 1.3) and I've tried really hard to do so until I found out that this test is incompatible with 1.3 and the data could not be generated with flink 1.3 Later found out the comment to the test class that also states so: 

> <p><b>Important:</b> Since Avro itself broke class compatibility between 1.7.7 (used in Flink 1.3) 

> * and 1.8.2 (used in Flink 1.4), the Avro by Pojo compatibility is broken through Avro already.
> * This test only tests that the Avro serializer change (switching from Pojo to Avro for Avro types)
> * works properly.

Also the commented code does not compile with flink 1.3(but this is a minor thing)

Data serialized with version of avro used in flink 1.3 (1.7.7) is not binary compatible with avro 1.8.2 (in flink 1.4+), due to changes how SpecificFixed is constructed.

Therefore how I regenerated this snapshot data is that I run the commented code on current branch. That is why I also changed few descriptions to that test as it test compatibility of `PojoSerializer` with `AvroSerializer` rather than binary backwards compatibility. 

Nevertheless I am more than happy to hear any comments on that.

dawidwys commented:
Also as for the package name or place where to put it, I don't feel competent to suggest a place, therefore will be happy to apply your suggestion.

StephanEwen commented:
I would actually keep the package name for now. It makes sense, because the connection to the registry is avro-specific at the moment...

dawidwys commented:
@StephanEwen could you have another look?

StephanEwen commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroDeserializationSchema.java:
```diff
@@ -0,0 +1,176 @@
+	AvroDeserializationSchema(Class<T> recordClazz, @Nullable Schema reader) {
+		Preconditions.checkNotNull(recordClazz, "Avro record class must not be null.");
+		this.recordClazz = recordClazz;
+		this.inputStream = new MutableByteArrayInputStream();
```
I would skip the initialization in the constructor, if you have he initialization in `checkAvroInitialized()`. Simpler, and avoids having two places that to the initialization which have to be kept in sync.

StephanEwen commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroDeserializationSchema.java:
```diff
@@ -0,0 +1,176 @@
+	 * @param schema schema of produced records
+	 * @return deserialized record in form of {@link GenericRecord}
+	 */
+	public static AvroDeserializationSchema<GenericRecord> forGeneric(Schema schema) {
```
Minor comment: I found it helps code structure/readability to move static/factory methods either to the top or the bottom of the class.

StephanEwen commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroDeserializationSchema.java:
```diff
@@ -0,0 +1,176 @@
+
+			return datumReader.read(null, decoder);
+		} catch (Exception e) {
+			throw new RuntimeException("Failed to deserialize message.", e);
```
Unnecessary exception wrapping, see other comment.

StephanEwen commented:
Added a few more comment, most importantly around exception wrapping.
Otherwise, looking good...

dawidwys commented:
I've addressed your comments @StephanEwen . If you don't have any more, I will merge it today.

StephanEwen commented:
Looks good, thanks!

+1 to merge this

medcv commented on flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/ConfluentRegistryAvroDeserializationSchema.java:
```diff
@@ -0,0 +1,136 @@
+	 * @param identityMapCapacity maximum number of cached schema versions (default: 1000)
+	 * @return deserialized record in form of {@link GenericRecord}
+	 */
+	public static ConfluentRegistryAvroDeserializationSchema<GenericRecord> forGeneric(Schema schema, String url,
```
@dawidwys couldn't this be `private`?

dawidwys replied:
End user is supposed to use only this or `forSpecific` method and no other one. Therefore it must be public.

medcv commented on flink-formats/flink-avro-confluent-registry/src/main/java/org/apache/flink/formats/avro/registry/confluent/ConfluentRegistryAvroDeserializationSchema.java:
```diff
@@ -0,0 +1,136 @@
+	 * @param identityMapCapacity maximum number of cached schema versions (default: 1000)
+	 * @return deserialized record
+	 */
+	public static <T extends SpecificRecord> ConfluentRegistryAvroDeserializationSchema<T> forSpecific(Class<T> tClass,
```
@dawidwys couldn't this be `private`?

cricket007 commented:
What about implementing a `KeyedDeserializationSchema` for Avro?
